Automattic/jetpack
# Make Contact Form error fixing accessible

Fixes https://github.com/Automattic/jetpack/issues/14155 https://github.com/Automattic/jetpack/issues/34235

## Proposed changes:

In this PR, we make Contact forms more accessible by overwriting the native form validation. Browsers handle form validation differently, making accessible forms for all users difficult. Specifically, we:
- Overwrite the form native validation by adding the `novalidate` attribute on the `form`.
- Register and enqueue the custom validation script.
- Rely on the [Contraint Validation API](https://developer.mozilla.org/en-US/docs/Learn/Forms/Form_validation#the_constraint_validation_api) to report fields validity.
- Show an error message above the submit button if one or more fields are invalid. This is the right place to display a message since the button is on the screen, and the user is focused on it. Scrolling or focusing on another place can be confusing. The error message is also read out loud by screen readers thanks to `aria-live=assertive`.
- For each invalid field, mark it as such and show a more specific message.

### Other information:


## Jetpack product discussion


n/a

## Does this pull request change what data or activity we track or use?


No.

## Testing instructions:






- Spin up a test site with this branch
- Create a new post
- Add a _Contact form_ block to the post
- View the preview
- Check that when one or more fields are invalid, an error message is displayed above the submit button and below each invalid field
- Fix the errors and make sure you can submit the form successfully
- Check that the layout is not broken for all three styles: _Default_, _Outlined_, and _Animated_
- Optionally, repeat these steps by adding various field types
- Optionally, test the above by adding a _Shortcode_ block instead
<img width="669" alt="Screenshot 2023-11-21 at 10 48 48 AM" src="https://github.com/Automattic/jetpack/assets/1620183/9776ca40-cca5-49fb-aa94-525779db6e09">

## Captures

### Before

Notice that the focus is immediately switched to the first invalid field on submitting. When tabbing out of the field, the error message disappears. This requires a user to submit the form again to be able to see the error message.

https://github.com/Automattic/jetpack/assets/1620183/1a8831f2-615e-4062-9c1d-702cfc421d69

### After

On submitting the form, an error message is shown above the submit button but no scroll or focusing is happening. We let the user tab back and forth to investigate which fields need to be corrected. Error messages are visible at all time and announced by the screen reader.

https://github.com/Automattic/jetpack/assets/1620183/5eb6bf92-db6a-4fa6-8607-667fcebb2d50
